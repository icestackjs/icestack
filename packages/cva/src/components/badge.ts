import { cva } from 'class-variance-authority'

import { typePrefix, expands, sizePrefix } from '@/shared'
import { InternalOptions } from '@/types'

export default function (opts: InternalOptions) {
  const { prefix } = opts
  const className = 'badge'
  const baseClass = prefix + className
  const basePrefix = baseClass + '-'

  const types = typePrefix(basePrefix)

  const sizes = sizePrefix(basePrefix)

  return {
    cva: cva([baseClass], {
      variants: {
        type: expands(types),
        outline: {
          true: basePrefix + 'outline'
        },
        size: expands(sizes)
      },
      defaultVariants: {}
    }),
    prefix,
    className,
    baseClass,
    types,
    sizes
  }
}
