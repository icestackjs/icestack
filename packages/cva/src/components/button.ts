import { cva } from 'class-variance-authority'

import { typePrefix, expands, addPrefix } from '@/shared'
import { InternalOptions } from '@/types'

export default function (opts: InternalOptions) {
  const { prefix } = opts
  const baseClass = prefix + 'btn'
  const basePrefix = baseClass + '-'

  const allTypes = typePrefix(basePrefix)

  const allSizes = addPrefix(basePrefix, ['xs', 'sm', 'md', 'lg', 'wide', 'block'])

  const allShapes = addPrefix(basePrefix, ['square', 'circle'])

  return {
    cva: cva([baseClass], {
      variants: {
        type: expands(allTypes),
        outline: {
          true: basePrefix + 'outline'
        },
        size: expands(allSizes),
        glass: {
          true: 'glass'
        },
        disabled: {
          true: basePrefix + 'disabled'
        },
        shape: expands(allShapes)
      },
      defaultVariants: {}
    }),
    baseClass,
    types: allTypes,
    sizes: allSizes,
    shapes: allShapes
  }
}
