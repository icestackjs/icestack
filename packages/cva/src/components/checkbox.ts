import { cva } from 'class-variance-authority'

import { typePrefix, expands, sizePrefix } from '@/shared'
import { InternalOptions } from '@/types'

export default function (opts: InternalOptions) {
  const { prefix } = opts
  const className = 'checkbox'
  const baseClass = prefix + className
  const basePrefix = baseClass + '-'
  const types = typePrefix(basePrefix)
  const sizes = sizePrefix(basePrefix)

  return {
    cva: cva([baseClass], {
      variants: {
        type: expands(types),
        size: expands(sizes)
      },
      defaultVariants: {}
    }),
    prefix,
    className,
    baseClass,
    types,
    sizes
  }
}
