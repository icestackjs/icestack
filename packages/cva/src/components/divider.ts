import { cva } from 'class-variance-authority'

import { addPrefix, expands, typePrefix } from '@/shared'
import { InternalOptions } from '@/types'

export default function (opts: InternalOptions) {
  const { prefix } = opts
  const className = 'divider'
  const baseClass = prefix + className
  const basePrefix = baseClass + '-'
  const types = typePrefix(basePrefix)
  const positions = addPrefix(basePrefix, ['vertical', 'horizontal', 'end', 'start'])

  return {
    cva: cva([baseClass], {
      variants: {
        type: expands(types),
        position: expands(positions)
      },
      defaultVariants: {}
    }),
    prefix,
    className,
    baseClass,
    positions,
    types
  }
}
