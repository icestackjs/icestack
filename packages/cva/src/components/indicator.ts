import { cva, VariantProps } from 'class-variance-authority'

import { addPrefix, expands } from '@/shared'
import { InternalOptions } from '@/types'

export default function (opts: InternalOptions) {
  const { prefix } = opts
  const className = 'indicator'
  const baseClass = prefix + className
  const basePrefix = baseClass + '-'

  const positions = addPrefix(basePrefix, ['start', 'center', 'end', 'top', 'middle', 'bottom'])
  const parent = cva([baseClass])

  const child = cva([basePrefix + 'item'], {
    variants: {
      position: expands(positions)
    },
    defaultVariants: {}
  })

  return {
    cva: {
      parent,
      child
    },
    prefix,
    className,
    baseClass,
    positions
  }
}
