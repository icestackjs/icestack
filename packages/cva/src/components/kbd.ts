import { cva } from 'class-variance-authority'

import { expands, sizePrefix } from '@/shared'
import { InternalOptions } from '@/types'

export default function (opts: InternalOptions) {
  const { prefix } = opts
  const className = 'kbd'
  const baseClass = prefix + className
  const basePrefix = baseClass + '-'
  const sizes = sizePrefix(basePrefix)
  return {
    cva: cva([baseClass], {
      variants: {
        size: expands(sizes)
      },
      defaultVariants: {}
    }),
    prefix,
    className,
    baseClass,
    sizes
  }
}
