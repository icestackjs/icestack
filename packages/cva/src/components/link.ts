import { cva } from 'class-variance-authority'

import { typePrefix, expands } from '@/shared'
import { InternalOptions } from '@/types'

export default function (opts: InternalOptions) {
  const { prefix } = opts
  const className = 'link'
  const baseClass = prefix + className
  const basePrefix = baseClass + '-'
  const types = typePrefix(basePrefix)
  return {
    cva: cva([baseClass], {
      variants: {
        type: expands(types),
        hover: {
          true: basePrefix + 'hover'
        }
      },
      defaultVariants: {}
    }),
    prefix,
    className,
    baseClass,
    types
  }
}
