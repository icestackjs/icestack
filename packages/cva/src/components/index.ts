// import type { VariantProps } from 'class-variance-authority'
import button from './button'
import loading from './loading'
import mask from './mask'
import { UserDefinedOptions } from '@/types'
import { getOptions } from '@/options'

export type { VariantProps } from 'class-variance-authority'

export function createCvaWithAddition(opts: UserDefinedOptions = {}) {
  const options = getOptions(opts)
  return {
    button: button(options),
    loading: loading(options),
    mask: mask(options)
  }
}

export function createCva(opts: UserDefinedOptions = {}) {
  const options = getOptions(opts)
  return {
    button: button(options).cva,
    loading: loading(options).cva,
    mask: mask(options).cva
  }
}
