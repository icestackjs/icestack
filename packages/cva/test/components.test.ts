import { createCva } from '@/components'

describe('components', () => {
  const CVA = createCva()

  it('all components default', () => {
    const map: Record<string, string> = {}
    for (const [name, obj] of Object.entries(CVA)) {
      if (typeof obj === 'function') {
        map[name] = obj()
      } else if (typeof obj === 'object') {
        // @ts-ignore
        map[name] = Object.entries(obj).reduce((acc, [name, fn]) => {
          // @ts-ignore
          acc[name] = fn()
          return acc
        }, {})
      }
    }
    expect(map).toMatchSnapshot()
  })
})
