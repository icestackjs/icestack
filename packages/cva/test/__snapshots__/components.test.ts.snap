// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components > all components default 1`] = `
{
  "alert": "alert",
  "badge": "badge",
  "button": "btn",
  "checkbox": "checkbox",
  "divider": "divider",
  "indicator": {
    "child": "indicator-item",
    "parent": "indicator",
  },
  "input": "input",
  "kbd": "kbd",
  "link": "link",
  "loading": "loading",
  "mask": "mask",
  "progress": "progress",
  "radio": "radio",
  "range": "range",
  "select": "select-",
  "skeleton": "skeleton",
  "steps": {
    "child": "step",
    "parent": "steps",
  },
  "tab": {
    "child": "tab",
    "parent": "tabs",
  },
  "table": "table",
  "textarea": "textarea",
  "toast": "toast",
  "toggle": "toggle",
  "tooltip": "tooltip",
}
`;
